### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by IRuban**

### CODE REVIEW

File: spec/features/pages/root_spec.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  scenario 'Visitor root path and click All books' do
+    visit root_path
+    find('li', class: 'dropdown', text: 'Shop').click
+    click_link('All books')
+    expect(page).to have_content 'Сatalog'
+    expect(page).to have_current_path '/books'
+  end
+
+  scenario 'Visitor root path and click slider book' do
+    books = FactoryBot.create(:book)
```
Comment (differencialx) [inline spec/features/pages/root_spec.rb]: use let for this in all tests

File: spec/features/pages/root_spec.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      fill_in 'user_password', with: user[:password]
+      fill_in 'user_password_confirmation', with: user[:password]
+    end
+    click_on('Sign up')
+    expect(page).to have_content 'Welcome! You have signed up successfully.'
+    expect(User.count).to eq(1)
+  end
+
+  scenario 'Visitor login page and click sing up not to be successfull' do
+    visit root_path
```
Comment (differencialx) [inline spec/features/pages/root_spec.rb]: I think you should move all sign_in, sign_up tests into separate file and check only that you was redirected to sign in or sign up page from root page

File: spec/models/book_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Book, type: :model do
+  let(:book) { FactoryBot.create :book }
+
+  context 'Column in db' do
+    it { is_expected.to have_db_column(:caption).of_type(:string) }
```
Comment (differencialx) [inline spec/models/book_spec.rb]: old syntax
https://gist.github.com/luisalima/4339490

File: spec/models/book_spec.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    it { is_expected.to have_db_column(:price).of_type(:float) }
+    it { is_expected.to have_db_column(:image).of_type(:string) }
+    it { is_expected.to have_db_column(:date_of_publication).of_type(:datetime) }
+    it { is_expected.to have_db_column(:height).of_type(:float) }
+    it { is_expected.to have_db_column(:width).of_type(:float) }
+    it { is_expected.to have_db_column(:depth).of_type(:float) }
+    it { is_expected.to have_db_column(:material).of_type(:string) }
+  end
+  context 'have assosiations' do
+    it { expect(book).to respond_to :authors }
```
Comment (differencialx) [inline spec/models/book_spec.rb]: use shoulda matchers

File: spec/models/category_spec.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Category, type: :model do
+  let(:category) { FactoryBot.create :category }
+
+  context 'Column in db' do
+    it { is_expected.to have_db_column(:name).of_type(:string) }
```
Comment (differencialx) [inline spec/models/category_spec.rb]: the same

File: app/controllers/pages_controller.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+class PagesController < ApplicationController
+  NUMBER_SLIDER = 3
+  NUMBER_BEST_SELLER = 4
+
+  decorates_assigned :slider_books
+
+  def home
+    @best_seller = Book.best_sellers
+    @slider_books = Book.first(NUMBER_SLIDER)
```
Comment (differencialx) [inline app/controllers/pages_controller.rb]: wrong

File: app/controllers/users/registrations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
```
Comment (differencialx) [inline app/controllers/users/registrations_controller.rb]: ########################

File: app/decorators/order_decorator.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+class OrderDecorator < Draper::Decorator
+  HIDDEN_NUMBERS = '**** **** ****'
+  delegate_all
+
+  def order_summary
+    line_items.sum(:item_price).round(2)
+  end
+
+  def sum_books
+    line_items.decorate.sum(&:quantity)
```
Comment (differencialx) [inline app/decorators/order_decorator.rb]: use sum method of active record

File: app/models/ability.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    guest_ability
```
Comment (differencialx) [inline app/models/ability.rb]: refactor

File: app/models/book.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class Book < ApplicationRecord
+  mount_uploader :image, ImageUploader
+
+  belongs_to :category
+  has_many :line_items
+  has_many :book_authors, dependent: :destroy
+  has_many :authors, through: :book_authors
```
Comment (differencialx) [inline app/models/book.rb]: remove

File: app/models/book.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  has_many :comments, dependent: :destroy
+  has_many :users, through: :comments
+  has_many :authors, through: :book_authors
+  has_many :book_images, dependent: :destroy
+
+  scope :by_category, ->(id) { where(category_id: id) }
+  scope :last_three, ->(quantity) { order(created_at: :desc).limit(quantity) }
```
Comment (differencialx) [inline app/models/book.rb]: rename to latest

File: app/models/category.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class Category < ApplicationRecord
+  has_many :books, dependent: :destroy
+
+  validates :name, presence: true
```
Comment (differencialx) [inline app/models/category.rb]: should be uniq

File: app/presenters/books_presenter.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class BooksPresenter
+  SORT = {
+    caption_asc: I18n.t('categories.caption_asc'),
+    caption_desc: I18n.t('categories.caption_desc'),
+    id_desc: I18n.t('categories.id_desc'),
```
Comment (differencialx) [inline app/presenters/books_presenter.rb]: created_at

File: app/services/checkout/update/confirm.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
```
Comment (differencialx) [inline app/services/checkout/update/confirm.rb]: ########################

File: app/models/order.rb [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  KEYS_FOR_STATUSES = %i[in_progress in_delivery delivered canceled].freeze
+
+  belongs_to :user, optional: true
+  has_many :line_items, dependent: :destroy
+  has_many :addresses, as: :addressable, dependent: :destroy
+  has_one :billing_address, as: :addressable, dependent: :destroy
+  has_one :shipping_address, as: :addressable, dependent: :destroy
+  has_one :payment, dependent: :destroy
+  belongs_to :delivery, optional: true
+  belongs_to :coupon, optional: true
```
Comment (differencialx) [inline app/models/order.rb]: add validation for number uniq

File: app/services/checkout/update/confirm.rb [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+  def call
+    check_generate_number_for_orders
+  end
+
+  private
+
+  def check_generate_number_for_orders
+    loop do
+      @number = SecureRandom.random_number.to_s.slice(FIRST_NUMBER, LAST_NUMBER)
```
Comment (differencialx) [inline app/services/checkout/update/confirm.rb]: use range

File: app/services/orders/check_order_service.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class Orders::CheckOrderService
+  def initialize(order_id, user)
+    @order_id = order_id
+    @user = user
+  end
+
+  STATUSES = %w[in_progress in_delivery delivered canceled].freeze
```
Comment (differencialx) [inline app/services/orders/check_order_service.rb]: duplication

File: config/initializers/carrierwave.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+    config.fog_credentials = {
+      provider: 'AWS',
+      aws_access_key_id: Rails.application.credentials.prod[:aws][:access_key_id],
+      aws_secret_access_key: Rails.application.credentials.prod[:aws][:secret_access_key],
+      region: 'us-east-2'
+    }
+    config.storage = :fog
+    config.fog_directory = 'bookstore-img'
+  else
+    config.storage = :file
```
Comment (differencialx) [inline config/initializers/carrierwave.rb]: configure for tests

File: db/migrate/20190518211041_create_coupons.rb [hunk 19]
```diff
@@ -0,0 +1,9 @@
+class CreateCoupons < ActiveRecord::Migration[5.2]
+  def change
+    create_table :coupons do |t|
+      t.string :key
```
Comment (differencialx) [inline db/migrate/20190518211041_create_coupons.rb]: add index

File: spec/controllers/addresses_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
```
Comment (differencialx) [inline spec/controllers/addresses_controller_spec.rb]: ###################################

File: spec/decorators/book_decorator_spec.rb [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+RSpec.describe BookDecorator do
+  let(:book) { build_stubbed(:book, :with_author).decorate }
+
+  context 'when description' do
+    let(:valid_short_description_size) { 100 }
+
+    it { expect(book.short_description.size).to eq valid_short_description_size }
+    it { expect(book.dimensions).to eq("H: #{book.width} x W: #{book.height} x D: #{book.depth}") }
+    it { expect(book.publish_date).to eq(book.date_of_publication.strftime('%d/%m/%Y')) }
+    it { expect(book.long_description?).to eq(true) }
```
Comment (differencialx) [inline spec/decorators/book_decorator_spec.rb]: be_truthy

File: spec/forms/address_form_spec.rb [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+    let(:user) { create(:user) }
+    let(:billing_address) { 'BillingAddress' }
+    let(:params) do
+      { first_name: valid_input, last_name: valid_input, address: valid_input, city: valid_input, zip: valid_zip,
+        country: valid_input, phone: valid_phone, type: billing_address }
+    end
+
+    it 'create billing address' do
+      expect(BillingAddress.count).to eq(0)
+      expect(subject.save_address(params, user)).to eq(true)
```
Comment (differencialx) [inline spec/forms/address_form_spec.rb]: be_truthy

File: spec/forms/address_form_spec.rb [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+                            zip: valid_zip, country: valid_input, phone: valid_phone, type: billing_address)
+    end
+    let(:params) do
+      { first_name: valid_input, last_name: valid_input, address: valid_input, city: valid_input, zip: valid_zip,
+        country: valid_input, phone: valid_phone, type: billing_address }
+    end
+
+    it 'update billing address' do
+      expect(BillingAddress.count).to eq(0)
+      expect(subject.save_address(params, user)).to eq(true)
```
Comment (differencialx) [inline spec/forms/address_form_spec.rb]: be_truthy

File: spec/decorators/book_decorator_spec.rb [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+    it { expect(book.full_description).to eq(book.description) }
+  end
+
+  context 'when authors_as_string' do
+    let(:book_author) { book.authors.first.to_s }
+
+    it { expect(book.authors_title.class).to eq String }
```
Comment (differencialx) [inline spec/decorators/book_decorator_spec.rb]: remove this, it's extra

File: spec/factories/billing_address.rb [hunk 25]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :billing_address do
+    first_name { Faker::FunnyName.two_word_name }
+    last_name { Faker::FunnyName.two_word_name }
+    address { Faker::FunnyName.two_word_name }
+    city { Faker::FunnyName.two_word_name }
+    zip { 20204 }
```
Comment (differencialx) [inline spec/factories/billing_address.rb]: use FFaker

File: spec/factories/billing_address.rb [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+    country { 'AX' }
```
Comment (differencialx) [inline spec/factories/billing_address.rb]: same

File: spec/factories/billing_address.rb [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+    phone { '+380636472129' }
```
Comment (differencialx) [inline spec/factories/billing_address.rb]: same

File: spec/factories/billing_address.rb [hunk 28]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :billing_address do
```
Comment (differencialx) [inline spec/factories/billing_address.rb]: you should have `addresses.rb` for all factories for addresses
don't split them in separate files

File: spec/factories/billing_address.rb [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+    first_name { Faker::FunnyName.two_word_name }
+    last_name { Faker::FunnyName.two_word_name }
+    address { Faker::FunnyName.two_word_name }
+    city { Faker::FunnyName.two_word_name }
+    zip { 20204 }
+    country { 'AX' }
+    phone { '+380636472129' }
+    type { 'BillingAddress' }
```
Comment (differencialx) [inline spec/factories/billing_address.rb]: it's extra, as `factory :billing_address` will use Billing address model which will set type field

File: app/models/coupon.rb [hunk 30]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class Coupon < ApplicationRecord
+  has_many :orders, dependent: :nullify
```
Comment (differencialx) [inline app/models/coupon.rb]: add validation for key uniqueness

File: spec/factories/coupon.rb [hunk 31]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (differencialx) [inline spec/factories/coupon.rb]: rename file to `coupons`

File: spec/factories/comment.rb [hunk 32]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (differencialx) [inline spec/factories/comment.rb]: factories files names should have plural naming, not a singular 
please check all factories file names

File: spec/factories/line_item.rb [hunk 33]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :line_item do
+    quantity { 1 }
+    book
+    item_price { quantity * Book.last.price }
```
Comment (differencialx) [inline spec/factories/line_item.rb]: move this logic to after create callback for factory

File: spec/factories/shipping_address.rb [hunk 34]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :shipping_address do
```
Comment (differencialx) [inline spec/factories/shipping_address.rb]: the same as for billing address

Comment (differencialx) [inline spec/factories/shipping_address.rb]: also move all these fields to address factory and use it as super class for billing and shipping address factories

**Comment by IRuban** [comment]
I think not by chance)

вс, 9 июня 2019 г. в 00:26, differencialx <notifications@github.com>:

File: spec/controllers/addresses_controller_spec.rb [hunk 35]
```diff
@@ -3,7 +3,6 @@
 require 'rails_helper'
 
 RSpec.describe AddressesController, type: :controller do
-  let(:user) { create(:user) }
```
Comment (differencialx) [inline spec/controllers/addresses_controller_spec.rb]: use shared context for user login

File: spec/factories/addresses.rb [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+  factory :address do
+    first_name { Faker::FunnyName.two_word_name }
+    last_name { Faker::FunnyName.two_word_name }
+    address { Faker::Address.street_address }
+    city { Faker::Address.city }
+    zip { Faker::Address.zip_code }
+    country { Faker::Address.country_code }
+    phone { '+380636472129' }
+
+    trait :shipping_address do
```
Comment (differencialx) [inline spec/factories/addresses.rb]: create separate factory

File: spec/rails_helper.rb [hunk 37]
```diff
@@ -16,6 +16,8 @@
 require 'bcrypt'
 require 'cancan/matchers'
 require 'rack_session_access/capybara'
+require 'pry-rails'
+require 'rspec/expectations'
```
Comment (differencialx) [inline spec/rails_helper.rb]: remove

**Pull request merged**